src/realm/group_shared.cpp

```diff
@@ -1860,6 +1859,8 @@ void SharedGroup::low_level_commit(uint_fast64_t new_version)
     out.set_versions(new_version, oldest_version);
     // Recursively write all changed arrays to end of file
     ref_type new_top_ref = out.write_group(); // Throws
+    m_free_space = out.get_free_space();
+    m_used_space = out.get_file_size() - m_free_space;
```

Comment:
The freelist updating still has some `O(N^2)` parts, so I rather hope `get_free_space()` doesn't have the same complexity.